### PR TITLE
Part 2 of 4: Auditing our required parameters for the C# SDK.

### LINKED ISSUE #25: Allow users to omit parameters for optional defaults
# Customer Issue

Many Smartsheet API calls allow for customisation of query options through a wide variety of optional properties. This can be shown by looking at one of the SDK interfaces for accessing Smartsheet functionality: retrieving a sheet.

From the `SheetResources` interface in [smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs)

```
        Sheet GetSheet(
                    long sheetId,
                    IEnumerable<SheetLevelInclusion> includes,
                    IEnumerable<SheetLevelExclusion> excludes,
                    IEnumerable<long> rowIds,
                    IEnumerable<int> rowNumbers,
                    IEnumerable<long> columnIds,
                    long? pageSize,
                    long? page); 
```

The documentation for the parameters from `includes` onwards, denotes that the properties are optional as part of the task of retrieving a sheet e.g. `/// <param name="includes"> used to specify the optional objects to include. </param>`

The parameters are optional for the underlying http API call, but not for users of the C# SDK, who have to provide an input for all parameters, bar the final two, `pageSize`, and `page`, which are denoted as optional owing to them being a nullable long i.e. `long?`.

This results in invocations to the `GetSheet` implementation like this from https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/sdk-csharp-sample-net60/Program.cs#L30:

```
    var sheet = smartsheet.SheetResources.GetSheet(sheetId, null, null, null, null, null, null, null);
```

# Proposal

A review of the interfaces throughout the API should take place to establish which of the parameters are optional, and which are mandatory. After identifying genuinely optional parameters, we can modify existing interfaces and implementations to mark parameters as optional using the [Nullable operator](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types) `?`.

We should then expect to see invocations to Smartsheet SDK functionality that omit parameters where they are the default:

```
    var sheet = smartsheet.SheetResources.GetSheet(sheetId);
```

# Next Steps

Interfaces should be reviewed to consider which methods contain parameters which are permittable by the Smartsheet API to be `null` values.

The full list of interfaces to review:

- [x] [AttachmentVersioningResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/AttachmentVersioningResources.cs#L29)
- [x] [CommentAttachmentResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/CommentAttachmentResources.cs#L30)
- [x] [ContactResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/ContactResources.cs#L30)
- [x] [DiscussionAttachmentResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/DiscussionAttachmentResources.cs#L30)
- [x] [DiscussionCommentResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/DiscussionCommentResources.cs#L28)
- [x] [EventResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/EventResources.cs#L23)
- [x] [FavoriteResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/FavoriteResources.cs#L32)
- [x] [FolderResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/FolderResources.cs#L30)
- [x] [FolderSheetResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/FolderSheetResources.cs#L30)
- [x] [GroupResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/GroupResources.cs#L30)
- [x] [HomeFolderResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/HomeFolderResources.cs#L30)
- [x] [HomeResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/HomeResources.cs#L31)
- [x] [ImageUrlsResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/ImageUrlsResources.cs#L33)
- [x] [PassthroughResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/PassthroughResources.cs#L26)
- [x] [ReportResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/ReportResources.cs#L30)
- [x] [RowAttachmentResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/RowAttachmentResources.cs#L30)
- [x] [RowColumnResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/RowColumnResources.cs#L30)
- [x] [RowDiscussionResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/RowDiscussionResources.cs#L30)
- [x] [SearchResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SearchResources.cs#L31)
- [x] [ShareResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/ShareResources.cs#L32)
- [x] [SheetAttachmentResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetAttachmentResources.cs#L29)
- [x] [SheetAutomationRuleResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetAutomationRuleResources.cs#L32)
- [x] [SheetColumnResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetColumnResources.cs#L31)
- [x] [SheetCommentResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetCommentResources.cs#L29)
- [x] [SheetCrossSheetReferenceResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetCrossSheetReferenceResources.cs#L27)
- [x] [SheetDiscussionResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetDiscussionResources.cs#L29)
- [x] [SheetFilterResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetFilterResources.cs#L33)
- [x] [SheetResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs#L33)
- [x] [SheetRowResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetRowResources.cs#L32)
- [x] [SheetSummaryResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetSummaryResources.cs#L30)
- [x] [SheetUpdateRequestResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetUpdateRequestResources.cs#L23)
- [x] [SightResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SightResources.cs#L26)
- [x] [TemplateResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/TemplateResources.cs#L30)
- [x] [TokenResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/TokenResources.cs#L30)
- [x] [UserResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/UserResources.cs#L30)
- [x] [UserSheetResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/UserSheetResources.cs#L31)
- [x] [WebhookResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/WebhookResources.cs#L28)
- [x] [WorkspaceFolderResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/WorkspaceFolderResources.cs#L30)
- [x] [WorkspaceResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/WorkspaceResources.cs#L31)
- [x] [WorkspaceSheetResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/WorkspaceSheetResources.cs#L30)

### PR DESCRIPTION
**Pull request by timothodge**
Looking to fix things up based on this open issue: https://github.com/smartsheet/smartsheet-csharp-sdk/issues/25

This is half of the audit done. Need to continue to do the other 1/2 but wanted to make sure people liked this before I move on. Here is the public API docs for anyone who wants to cross check: https://smartsheet.redoc.ly/tag/groups/

### CODE REVIEW

File: smartsheet-csharp-sdk/main/Smartsheet/Api/HomeFolderResources.cs [hunk 1]
```diff
@@ -48,7 +48,7 @@ public interface HomeFolderResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Folder> ListFolders(PaginationParameters paging);
+        PaginatedResult<Folder> ListFolders(PaginationParameters? paging = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/HomeFolderResources.cs:51]: Screenshot 2024-04-30 at 1 43 47 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/ReportResources.cs [hunk 2]
```diff
@@ -70,7 +70,7 @@ public interface ReportResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Report GetReport(long reportId, IEnumerable<ReportInclusion> include, int? pageSize, int? page, int? level);
+        Report GetReport(long reportId, IEnumerable<ReportInclusion>? include = null, int? pageSize = null, int? page = null, int? level = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/ReportResources.cs:73]: Screenshot 2024-04-30 at 2 18 18 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/ReportResources.cs [hunk 3]
```diff
@@ -91,7 +91,7 @@ public interface ReportResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Report> ListReports(PaginationParameters paging, DateTime? modifiedSince = null);
+        PaginatedResult<Report> ListReports(PaginationParameters? paging = null, DateTime? modifiedSince = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/ReportResources.cs:94]: Screenshot 2024-04-30 at 2 18 53 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/RowAttachmentResources.cs [hunk 4]
```diff
@@ -44,7 +44,7 @@ public interface RowAttachmentResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Attachment> ListAttachments(long sheetId, long rowId, PaginationParameters paging);
+        PaginatedResult<Attachment> ListAttachments(long sheetId, long rowId, PaginationParameters? paging = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/RowAttachmentResources.cs:47]: Screenshot 2024-04-30 at 2 27 28 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/RowAttachmentResources.cs [hunk 5]
```diff
@@ -64,7 +64,7 @@ public interface RowAttachmentResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Attachment AttachFile(long sheetId, long rowId, string file, string fileType);
+        Attachment AttachFile(long sheetId, long rowId, string file, string? fileType = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/RowAttachmentResources.cs:67]: If file type is null it defaults to application/octet-stream in the impl class.

File: smartsheet-csharp-sdk/main/Smartsheet/Api/RowColumnResources.cs [hunk 6]
```diff
@@ -47,7 +47,7 @@ public interface RowColumnResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<CellHistory> GetCellHistory(long sheetId, long rowId, long columnId, IEnumerable<CellInclusion> include, PaginationParameters paging);
+        PaginatedResult<CellHistory> GetCellHistory(long sheetId, long rowId, long columnId, IEnumerable<CellInclusion>? include = null, PaginationParameters? paging = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/RowColumnResources.cs:50]: Screenshot 2024-04-30 at 2 37 56 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/RowColumnResources.cs [hunk 7]
```diff
@@ -86,7 +86,7 @@ public interface RowColumnResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        void AddImageToCell(long sheetId, long rowId, long columnId, string file, string fileType);
+        void AddImageToCell(long sheetId, long rowId, long columnId, string file, string? fileType = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/RowColumnResources.cs:89]: Screenshot 2024-04-30 at 2 38 29 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/RowDiscussionResources.cs [hunk 8]
```diff
@@ -81,7 +81,7 @@ public interface RowDiscussionResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Discussion> ListDiscussions(long sheetId, long rowId, IEnumerable<DiscussionInclusion> include, PaginationParameters paging);
+        PaginatedResult<Discussion> ListDiscussions(long sheetId, long rowId, IEnumerable<DiscussionInclusion>? include = null, PaginationParameters? paging = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/RowDiscussionResources.cs:84]: Screenshot 2024-04-30 at 2 42 34 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SearchResources.cs [hunk 9]
```diff
@@ -61,8 +61,8 @@ public interface SearchResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        SearchResult Search(string query, IEnumerable<SearchInclusion> includes, SearchLocation? location,
-            DateTime? modifiedSince, IEnumerable<SearchScope> scopes);
+        SearchResult Search(string query, IEnumerable<SearchInclusion>? includes = null, SearchLocation? location = null,
+            DateTime? modifiedSince = null, IEnumerable<SearchScope>? scopes = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SearchResources.cs:65]: Screenshot 2024-04-30 at 2 46 39 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/ShareResources.cs [hunk 10]
```diff
@@ -48,7 +48,7 @@ public interface ShareResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Share> ListShares(long objectId, PaginationParameters paging, ShareScope shareScope);
+        PaginatedResult<Share> ListShares(long objectId, PaginationParameters? paging = null, ShareScope? shareScope = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/ShareResources.cs:51]: Screenshot 2024-04-30 at 2 52 11 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/ShareResources.cs [hunk 11]
```diff
@@ -89,7 +89,7 @@ public interface ShareResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        IList<Share> ShareTo(long objectId, IEnumerable<Share> shares, bool? sendEmail);
+        IList<Share> ShareTo(long objectId, IEnumerable<Share> shares, bool? sendEmail = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/ShareResources.cs:92]: Screenshot 2024-04-30 at 2 52 45 PM

**Review by SmarMykel — DISMISSED** [review]
Thanks for adding the tests. Looks good!

File: smartsheet-csharp-sdk/main/Smartsheet/Api/HomeResources.cs [hunk 12]
```diff
@@ -67,7 +67,7 @@ public interface HomeResources
         /// <exception cref="SmartsheetException">if any other error occurred during the operation</exception>
         /// <returns> the resource (note that if there is no such resource, this method will throw ResourceNotFoundException
         /// rather than returning null). </returns>
-        PersonalFolder GetFoldersPersonal(IEnumerable<HomeInclusion> includes, IEnumerable<HomeExclusion> excludes);
+        PersonalFolder GetFoldersPersonal(IEnumerable<HomeInclusion>? includes = null, IEnumerable<HomeExclusion>? excludes= null);
```
Comment (rzaracota) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/HomeResources.cs]: nitpick: spacing
```suggestion
        PersonalFolder GetFoldersPersonal(IEnumerable<HomeInclusion>? includes = null, IEnumerable<HomeExclusion>? excludes = null);
```
#

**Review by rzaracota — DISMISSED** [review]

File: smartsheet-csharp-sdk/main/Smartsheet/Api/HomeResources.cs (hunk 12 above)
Comment (timothodge) [reply smartsheet-csharp-sdk/main/Smartsheet/Api/HomeResources.cs]: Fixed it up in the latest commit thank you!

**Review by SmarMykel — APPROVED** [review]

**Pull request merged**